common/core/desktop/tests/unit/kmx/kmx_key_list.cpp

```diff
@@ -0,0 +1,77 @@
+/*
+  Copyright:    © 2018 SIL International.
+  Description:  Tests for key rules list generation in the kmx processor
+  Create Date:  30 Oct 2018
+  Authors:      Ross Cruickshank (RC) Marc Durdin (MD), Tim Eves (TSE)
+
+*/
+
+#include <kmx/kmx_processevent.h>
+
+#include "path.hpp"
+#include "state.hpp"
+
+#include "../test_assert.h"
+#include "../test_color.h"
+
+using namespace km::kbp::kmx;
+
+km_kbp_option_item test_env_opts[] =
+{
+  KM_KBP_OPTIONS_END
+};
+
+int error_args() {
+    std::cerr << "kmx: Not enough arguments." << std::endl;
+    return 1;
+}
+
+void test_key_list(const km::kbp::path &source_path){
+
+  km_kbp_keyboard * test_kb = nullptr;
+  km_kbp_state * test_state = nullptr;
+  km_kbp_keyboard_key * kb_key_list;
+
+  km::kbp::path const source_keybard = "039 - generic ctrlalt.kmx";
+  km::kbp::path full_path = km::kbp::path::join(source_path, source_keybard);
+
+  try_status(km_kbp_keyboard_load(full_path.native().c_str(), &test_kb));
+
+  // Setup state, environment
+  try_status(km_kbp_state_create(test_kb, test_env_opts, &test_state));
+
+  try_status(km_kbp_keyboard_get_key_list(test_kb,&kb_key_list));
+
+  km_kbp_keyboard_key *key_rule_it = kb_key_list;
+  auto n = 0;
+  for (; key_rule_it->key; ++key_rule_it) {
+    ++n;
+  }
```

Comment:
Can we assert the values returned as well? I would also like to see a test for a more complex keyboard with multiple groups and with the same keys referred in multiple groups (to validate that we don't list keys more than once). We cannot assume the list order so will need to be careful on testing results!